src/Model/LazyBSONDocument.php

```diff
@@ -0,0 +1,235 @@
+use function MongoDB\recursive_copy;
+use function sprintf;
+use function trigger_error;
+
+use const E_USER_WARNING;
+
+/**
+ * Model class for a BSON document.
+ *
+ * The internal data will be cast to an object during BSON serialization to
+ * ensure that it becomes a BSON document.
+ *
+ * @template TValue
+ * @template-implements ArrayAccess<string, TValue>
+ * @template-implements IteratorAggregate<string, TValue>
+ */
+class LazyBSONDocument implements ArrayAccess, IteratorAggregate
+{
+    /** @var Document<TValue> */
+    private Document $bson;
+
+    /** @var array<string, TValue> */
+    private array $read = [];
+
+    /** @var array<string, bool> */
+    private array $exists = [];
+
+    /** @var array<string, TValue> */
+    private array $set = [];
+
+    /** @var array<string, true> */
+    private array $unset = [];
+
+    private CodecLibrary $codecLibrary;
+
+    /**
+     * Deep clone this lazy document.
+     */
+    public function __clone()
+    {
+        $this->bson = clone $this->bson;
+
+        foreach ($this->set as $key => $value) {
+            $this->set[$key] = recursive_copy($value);
+        }
+    }
+
+    /**
+     * Constructs a lazy BSON document.
+     *
+     * @param Document<TValue>|array<TValue>|object|null $input An input for a lazy object.
+     *        When given a BSON document, this is treated as input. For arrays
+     *        and objects this constructs a new BSON document using fromPHP.
+     */
+    public function __construct($input = null, ?CodecLibrary $codecLibrary = null)
+    {
+        if ($input === null) {
+            $this->bson = Document::fromPHP([]);
+        } elseif ($input instanceof Document) {
+            $this->bson = $input;
+        } elseif (is_array($input) || is_object($input)) {
+            $this->bson = Document::fromPHP([]);
+
+            foreach ($input as $key => $value) {
+                assert(is_string($key));
+                $this->set[$key] = $value;
+                $this->exists[$key] = true;
+            }
+        } else {
+            throw InvalidArgumentException::invalidType('input', $input, [Document::class, 'array', 'null']);
+        }
+
+        $this->codecLibrary = $codecLibrary ?? new LazyBSONCodecLibrary();
+    }
+
+    /** @return TValue */
+    public function __get(string $property)
+    {
+        $this->readFromBson($property);
+
+        if (isset($this->unset[$property]) || ! $this->exists[$property]) {
+            trigger_error(sprintf('Undefined property: %s', $property), E_USER_WARNING);
+
+            return null;
+        }
+
+        return array_key_exists($property, $this->set) ? $this->set[$property] : $this->read[$property];
+    }
+
+    public function __isset(string $name): bool
+    {
+        // If we've looked for the value, return the cached result
+        if (isset($this->exists[$name])) {
+            return $this->exists[$name];
+        }
+
+        return $this->exists[$name] = $this->bson->has($name);
```

Comment:
```suggestion
        return $this->exists[$name] ??= $this->bson->has($name);
```

Reply:
I remember @jmikola commenting on a previous use of the null-coalesce operator, but don't exactly remember the circumstances. I've changed this here and in `LazyBSONArray`, but open to changing it back. I'll link the article that was previously linked in this context: https://localheinz.com/articles/2023/03/18/avoiding-one-liners-in-php/

Reply:
Honestly, I thought of that argument while writing the code suggestion. And since I really appreciate being able to rely on code coverage when improving test coverage, let's stick with this first version.